### DISCUSSION TITLE
How can you make user_nl_cam overwrite cam/cime_config/config_component.xml

### DISCUSSION THREAD

**Discussion by oyvindseland** [NorESMhub/CAM#80] [General]
I am trying to overwrite a namelist value that is defined in cam/cime_config/config_component.xml 
      <value compset="1850_CAM60%NORESM%NORBC" grid="a%1.9x2.5" > ocean_filename
='dms-hamocc-dow-taylor_chlor_a-lanaclim_N1850_f19_tn14_20190621_1751-1780_cycle
_version20190726.nc' </value>

with a user_nl value

&oslo_ctl_nl
ocean_filename		= 'dms-hamocc-dow-taylor_chlor_lanaclim_N1850_f19_tn14_F
orces_230512_1751-1780_cycle_version20230512.nc'
/

Config_components however seem to always overwrite the namelist.
Must I change the compset definitions to be able to change the namelist?  @gold2718

**Comment by gold2718** [comment]
Can you give me a git tag or a case directory to look at?

**Comment by oyvindseland** [comment]
On Betzy: 
Case: /cluster/projects/nn2345k/oyvinds/NorESM_Forces/cases/NF1850norbc_f19_f19_Forces_230703
Code: /cluster/projects/nn2345k/oyvinds/NorESM_Forces/NorESM

**Comment by gold2718** [comment]
I think you have some typos and unnecessary syntax in your user_nl_cam
Both `&clubb_params_nl` and `&micro_mg_nl` are not terminated so the namelist generator probably thinks you are trying to set `ocean_filename` as part of  `&clubb_params_nl`. However, you do not even need those so I recommend removing those lines as well as `&oslo_ctl_nl` and try `./preview_namelists` again.

**Comment by oyvindseland** [comment]
Yes, although I do not think these are critical when it is only one variable per namelist class.

At any rate I tried to build with the user_nl_cam only including
ocean_filename		= 'dms-hamocc-dow-taylor_chlor_lanaclim_N1850_f19_tn14_Forces_230512_1751-1780_cycle_version20230512.nc'

and it is still overwritten when using ./preview_namelists.       I have also tested ./check_case and with and without case.setup --clean.

The namelist variables I removed reverted to their default value, so user_nl_cam was read.

**Reply by gold2718** [reply]
> Yes, although I do not think these are critical when it is only one variable per namelist class.
> 
I think this is true when using a namelist with Fortran, I'm not so sure when dealing with the 'exceptional' logic of the CAM namelist build system.
I don't see anything wrong with this attempt although there is some unique logic in build_namelist at line 2892. Can you send the output of `./preview_namelists`? It would also be interesting to see if that `unless` clause is hit (which you could see by putting a `print` statement in there. Maybe something like:
```
if (defined $nl->get_value('ocean_filename')) {
    print "ocean_filename defined as $nl->get_value('ocean_filename')$eol";
} else {
    add_default($nl, 'ocean_filename');
    print "default value for ocean_filename set as $nl->get_value('ocean_filename')$eol";
}
```

**Reply by oyvindseland** [reply]
Output limited so just including it below.

The suggested print statements should go into the namelist build itself. Fails within the preview_namelists

./preview_namelists
Setting resource.RLIMIT_STACK to -1 from (8388608, -1)
Creating component namelists
   Calling /cluster/projects/nn2345k/oyvinds/NorESM_Forces/NorESM/components/cam//cime_config/buildnml
     ...calling cam buildcpp to set build time options
CAM namelist copy: file1 /cluster/projects/nn2345k/oyvinds/NorESM_Forces/cases/NF1850norbc_f19_f19_Forces_230703/Buildconf/camconf/atm_in file2 /cluster/work/users/oyvinds/noresm/NF1850norbc_f19_f19_Forces_230703/run/atm_in 
   Calling /cluster/projects/nn2345k/oyvinds/NorESM_Forces/NorESM/components/clm//cime_config/buildnml
   Calling /cluster/projects/nn2345k/oyvinds/NorESM_Forces/NorESM/components/cice//cime_config/buildnml
     ...calling cice buildcpp to set build time options
   Calling /cluster/projects/nn2345k/oyvinds/NorESM_Forces/NorESM/cime/src/components/data_comps/docn/cime_config/buildnml
   Calling /cluster/projects/nn2345k/oyvinds/NorESM_Forces/NorESM/components/mosart//cime_config/buildnml
   Calling /cluster/projects/nn2345k/oyvinds/NorESM_Forces/NorESM/cime/src/components/stub_comps/sglc/cime_config/buildnml
   Calling /cluster/projects/nn2345k/oyvinds/NorESM_Forces/NorESM/cime/src/components/stub_comps/swav/cime_config/buildnml
   Calling /cluster/projects/nn2345k/oyvinds/NorESM_Forces/NorESM/cime/src/components/stub_comps/sesp/cime_config/buildnml
   Calling /cluster/projects/nn2345k/oyvinds/NorESM_Forces/NorESM/cime/src/drivers/mct/cime_config/buildnml
   NOTE: ignoring setting of rof2ocn_liq_rmapname=idmap in seq_maps.rc
   NOTE: ignoring setting of rof2ocn_ice_rmapname=idmap in seq_maps.rc
Finished creating component namelists
